Airbnb Assignment by Sir Tariq

Student: Muhammad Ammar Jamshed 

In [147]:
import pandas as pd
import numpy as np
import seaborn as sns
import re
import string
from scipy import stats

In [148]:
df = pd.read_csv(r'C:\Users\muham\Downloads\singapairbnb_dirty_data .csv')

In [149]:
# first lets evaulate the data in a broad manner 
df

id                                               name    host_id  \
0        49091                  COZICOMFORT LONG TERM STAY ROOM 2     266763   
1        50646                    Pleasant Room along Bukit Timah     227796   
2        56334                                        COZICOMFORT     266763   
3        71609                Ensuite Room (Room 1 & 2) near EXPO     367042   
4        71896                    B&B  Room 1 near Airport & EXPO     367042   
...        ...                                                ...        ...   
7902  38105126  Loft 2 pax near Haw Par / Pasir Panjang. Free ...  278109833   
7903  38108273                         3bedroom luxury at Orchard  238891646   
7904  38109336    [ Farrer Park ] New City Fringe CBD Mins to MRT  281448565   
7905  38110493          Cheap Master Room in Central of Singapore  243835202   
7906  38112762  Amazing room with private bathroom walk to Orc...   28788520   

      host_name neighbourhood_group neighbourhood  latitude  longitude  \
0     Francesca        North Region     Woodlands   1.44255  103.79580   
1       Sujatha      Central Region   Bukit Timah   1.33235  103.78521   
2     Francesca        North Region     Woodlands   1.44246  103.79667   
3       Belinda         East Region      Tampines   1.34541  103.95712   
4       Belinda         East Region      Tampines   1.34567  103.95963   
...         ...                 ...           ...       ...        ...   
7902      Belle      Central Region    Queenstown   1.27973  103.78751   
7903       Neha      Central Region       Tanglin   1.29269  103.82623   
7904      Mindy      Central Region       Kallang   1.31286  103.85996   
7905      Huang      Central Region  River Valley   1.29543  103.83801   
7906    Terence      Central Region  River Valley   1.29672  103.83325   

            room_type  price minimum_nights  number_of_reviews last_review  \
0                   1     83            180                  1  21/10/2013   
1                   2     81             90                 18  26/12/2014   
2                   3     69              6                 20   1/10/2015   
3                   4    206              1                 14   11/8/2019   
4        Private room     94              1                 22  28/07/2019   
...               ...    ...            ...                ...         ...   
7902  Entire home/apt    100            NaN                  0         NaN   
7903  Entire home/apt    550            NaN                  0         NaN   
7904     Private room     58            NaN                  0         NaN   
7905     Private room     56            NaN                  0         NaN   
7906     Private room   1500            NaN                  0         NaN   

      reviews_per_month  calculated_host_listings_count  availability_365  
0                  0.01                               2             365.0  
1                  0.28                               1             365.0  
2                  0.20                               2             365.0  
3                  0.15                               9             353.0  
4                  0.22                               9             355.0  
...                 ...                             ...               ...  
7902                NaN                              31              61.0  
7903                NaN                              34             365.0  
7904                NaN                               3             173.0  
7905                NaN                               2              30.0  
7906                NaN                               7             365.0  

[7907 rows x 16 columns]

After evaulating the data set we can conclude that the data belongs to the Hoteling industry and is recording information
on customers who are using Airbnb to spend nights in hotels and lodges in Singapore

In [150]:
# lets see all the data types present in the data set and understand the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7907 entries, 0 to 7906
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              7895 non-null   object 
 1   name                            7863 non-null   object 
 2   host_id                         7907 non-null   object 
 3   host_name                       7907 non-null   object 
 4   neighbourhood_group             7841 non-null   object 
 5   neighbourhood                   7870 non-null   object 
 6   latitude                        7907 non-null   float64
 7   longitude                       7907 non-null   float64
 8   room_type                       7890 non-null   object 
 9   price                           7907 non-null   int64  
 10  minimum_nights                  7891 non-null   object 
 11  number_of_reviews               7907 non-null   int64  
 12  last_review                     51

In [151]:
#lets verify if there are null values in the data set 
df.isnull().sum()

id                                  12
name                                44
host_id                              0
host_name                            0
neighbourhood_group                 66
neighbourhood                       37
latitude                             0
longitude                            0
room_type                           17
price                                0
minimum_nights                      16
number_of_reviews                    0
last_review                       2758
reviews_per_month                 2758
calculated_host_listings_count       0
availability_365                   386
dtype: int64

# what is our strategy for filling missing values in the neighborhood?

In [152]:
# we can use regex wild cards in python to look for unique values by using unique identifying chars
print('column name' + '\t\t\t' + 'IsUnique')
df.apply(lambda x: x.is_unique,axis=0)

column name			IsUnique


id                                False
name                              False
host_id                           False
host_name                         False
neighbourhood_group               False
neighbourhood                     False
latitude                          False
longitude                         False
room_type                         False
price                             False
minimum_nights                    False
number_of_reviews                 False
last_review                       False
reviews_per_month                 False
calculated_host_listings_count    False
availability_365                  False
dtype: bool

In [153]:
#first we look for the null values in neighbour hood columns in which places
print(df[df['neighbourhood_group'].isnull()]['neighbourhood_group'].value_counts())

Series([], Name: neighbourhood_group, dtype: int64)


In [154]:
# we shall drop id since its not unique 
df = df.drop('id', axis=1)

In [155]:
# improvement, since we dropped one column
df.apply(lambda x: sum(x.isnull()),axis=0)

name                                44
host_id                              0
host_name                            0
neighbourhood_group                 66
neighbourhood                       37
latitude                             0
longitude                            0
room_type                           17
price                                0
minimum_nights                      16
number_of_reviews                    0
last_review                       2758
reviews_per_month                 2758
calculated_host_listings_count       0
availability_365                   386
dtype: int64

In [156]:
#first we look for the null values in neighbour hood columns, in which places
print(df[df['neighbourhood'].isnull()]['neighbourhood_group'].value_counts())

Central Region       31
East Region           3
North-East Region     2
North Region          1
Name: neighbourhood_group, dtype: int64


In [157]:
# we visualize the file here for ease of view
df

name    host_id  host_name  \
0                     COZICOMFORT LONG TERM STAY ROOM 2     266763  Francesca   
1                       Pleasant Room along Bukit Timah     227796    Sujatha   
2                                           COZICOMFORT     266763  Francesca   
3                   Ensuite Room (Room 1 & 2) near EXPO     367042    Belinda   
4                       B&B  Room 1 near Airport & EXPO     367042    Belinda   
...                                                 ...        ...        ...   
7902  Loft 2 pax near Haw Par / Pasir Panjang. Free ...  278109833      Belle   
7903                         3bedroom luxury at Orchard  238891646       Neha   
7904    [ Farrer Park ] New City Fringe CBD Mins to MRT  281448565      Mindy   
7905          Cheap Master Room in Central of Singapore  243835202      Huang   
7906  Amazing room with private bathroom walk to Orc...   28788520    Terence   

     neighbourhood_group neighbourhood  latitude  longitude        room_type  \
0           North Region     Woodlands   1.44255  103.79580                1   
1         Central Region   Bukit Timah   1.33235  103.78521                2   
2           North Region     Woodlands   1.44246  103.79667                3   
3            East Region      Tampines   1.34541  103.95712                4   
4            East Region      Tampines   1.34567  103.95963     Private room   
...                  ...           ...       ...        ...              ...   
7902      Central Region    Queenstown   1.27973  103.78751  Entire home/apt   
7903      Central Region       Tanglin   1.29269  103.82623  Entire home/apt   
7904      Central Region       Kallang   1.31286  103.85996     Private room   
7905      Central Region  River Valley   1.29543  103.83801     Private room   
7906      Central Region  River Valley   1.29672  103.83325     Private room   

      price minimum_nights  number_of_reviews last_review  reviews_per_month  \
0        83            180                  1  21/10/2013               0.01   
1        81             90                 18  26/12/2014               0.28   
2        69              6                 20   1/10/2015               0.20   
3       206              1                 14   11/8/2019               0.15   
4        94              1                 22  28/07/2019               0.22   
...     ...            ...                ...         ...                ...   
7902    100            NaN                  0         NaN                NaN   
7903    550            NaN                  0         NaN                NaN   
7904     58            NaN                  0         NaN                NaN   
7905     56            NaN                  0         NaN                NaN   
7906   1500            NaN                  0         NaN                NaN   

      calculated_host_listings_count  availability_365  
0                                  2             365.0  
1                                  1             365.0  
2                                  2             365.0  
3                                  9             353.0  
4                                  9             355.0  
...                              ...               ...  
7902                              31              61.0  
7903                              34             365.0  
7904                               3             173.0  
7905                               2              30.0  
7906                               7             365.0  

[7907 rows x 15 columns]

In [158]:
#thus for the missing reigons we can identify them with the particular city the affliated with the most
print(df[df['neighbourhood_group'].isnull()]['neighbourhood'].value_counts())

Geylang             12
River Valley         5
Tanglin              4
Bukit Merah          4
Novena               4
Rochor               4
Bukit Timah          3
Outram               3
Kallang              3
Queenstown           3
Downtown Core        2
Toa Payoh            2
Bedok                2
Yishun               1
Newton               1
Southern Islands     1
Orchard              1
Bukit Panjang        1
Clementi             1
Sembawang            1
Singapore River      1
Hougang              1
Serangoon            1
Bukit Batok          1
Ang Mo Kio           1
Jurong West          1
Choa Chu Kang        1
Bishan               1
Name: neighbourhood, dtype: int64


In [159]:
# lets now check if the missing values of 'River Valley' has been filled
print(df[df['neighbourhood_group'].isnull()]['neighbourhood'].value_counts())

Geylang             12
River Valley         5
Tanglin              4
Bukit Merah          4
Novena               4
Rochor               4
Bukit Timah          3
Outram               3
Kallang              3
Queenstown           3
Downtown Core        2
Toa Payoh            2
Bedok                2
Yishun               1
Newton               1
Southern Islands     1
Orchard              1
Bukit Panjang        1
Clementi             1
Sembawang            1
Singapore River      1
Hougang              1
Serangoon            1
Bukit Batok          1
Ang Mo Kio           1
Jurong West          1
Choa Chu Kang        1
Bishan               1
Name: neighbourhood, dtype: int64


In [160]:
# lets examine if there are other missing values in Central reigion 
print(df[df['neighbourhood_group'] == 'Central Region']['neighbourhood'].value_counts())

Kallang             1034
Geylang              972
Rochor               531
Novena               529
Outram               472
Bukit Merah          463
Downtown Core        426
River Valley         357
Queenstown           260
Tanglin              206
Singapore River      174
Marine Parade        171
Orchard              135
Newton               133
Bukit Timah          127
Toa Payoh             99
Museum                62
Bishan                56
Southern Islands      16
Marina South           1
Name: neighbourhood, dtype: int64


In [161]:
# we now replace the central reigon vallues with kallang as its the MODE
cond = (df['neighbourhood_group'] == 'Central Region') & (df['neighbourhood'].isnull())
df.loc[cond, 'neighbourhood'] = 'Kallang'

In [162]:
# Now we shall check if it has been filled 
print(df[df['neighbourhood'].isnull()]['neighbourhood_group'].value_counts())

East Region          3
North-East Region    2
North Region         1
Name: neighbourhood_group, dtype: int64


Thus its succeeded here now we have ti fill the same for East, North and North East by checking their mode 

In [163]:
# we check the mode of 'East Region'
print(df[df['neighbourhood_group'] == 'East Region']['neighbourhood'].value_counts())
print(df[df['neighbourhood_group'] == 'North-East Region']['neighbourhood'].value_counts())
print(df[df['neighbourhood_group'] == 'North Region']['neighbourhood'].value_counts())

Bedok        369
Pasir Ris     71
Tampines      63
Name: neighbourhood, dtype: int64
Hougang       108
Serangoon      68
Sengkang       67
Ang Mo Kio     56
Punggol        42
Name: neighbourhood, dtype: int64
Woodlands                  66
Yishun                     52
Sembawang                  40
Central Water Catchment    34
Sungei Kadut                5
Mandai                      3
Lim Chu Kang                1
Name: neighbourhood, dtype: int64


In [164]:
# we replace the value of the missing values of East Region with Bedok
cond1 = (df['neighbourhood_group'] == 'East Region') & (df['neighbourhood'].isnull())
df.loc[cond1, 'neighbourhood'] = 'Bedok'
# we over here replace the value of the missing values of the North-East region with Hougang
cond2 = (df['neighbourhood_group'] == 'North-East Region') & (df['neighbourhood'].isnull())
df.loc[cond2, 'neighbourhood'] = 'Hougang'
# we over here replace the value of the missing values of the North Region with Woodlands
cond3 = (df['neighbourhood_group'] == 'North Region') & (df['neighbourhood'].isnull())
df.loc[cond3, 'neighbourhood'] = 'Woodlands'

In [165]:
# thus we shall now check if all missing value replacement was a success
print(df[df['neighbourhood'].isnull()]['neighbourhood_group'].value_counts())

Series([], Name: neighbourhood_group, dtype: int64)


In [166]:
df.isnull().sum()

name                                44
host_id                              0
host_name                            0
neighbourhood_group                 66
neighbourhood                        0
latitude                             0
longitude                            0
room_type                           17
price                                0
minimum_nights                      16
number_of_reviews                    0
last_review                       2758
reviews_per_month                 2758
calculated_host_listings_count       0
availability_365                   386
dtype: int64

thus proven we have filled all null values in the Neighbourhood column by using the mode of each Region

#  what is our strategy for filling missing values in the neighborhood group?

In [167]:
print(df[df['neighbourhood_group'].isnull()]['neighbourhood'].value_counts())

Geylang             12
River Valley         5
Tanglin              4
Bukit Merah          4
Novena               4
Rochor               4
Bukit Timah          3
Outram               3
Kallang              3
Queenstown           3
Downtown Core        2
Toa Payoh            2
Bedok                2
Yishun               1
Newton               1
Southern Islands     1
Orchard              1
Bukit Panjang        1
Clementi             1
Sembawang            1
Singapore River      1
Hougang              1
Serangoon            1
Bukit Batok          1
Ang Mo Kio           1
Jurong West          1
Choa Chu Kang        1
Bishan               1
Name: neighbourhood, dtype: int64


In [168]:
df['neighbourhood_group'].value_counts()

Central Region       6255
West Region           535
East Region           506
North-East Region     343
North Region          202
Name: neighbourhood_group, dtype: int64

In [169]:
# we shall use a function with Regex characters in re.match method to fill the null values of Neighbourhood groups as they cant be fecthed
# using common filling methods due to the presence of a large no of unique values as shown in value counts
d = df['neighbourhood'].unique()
for val in d:
    i = str(df[df['neighbourhood'] == val]['neighbourhood_group'].value_counts())
    c = re.match('\w+(?:-\w+)+\s.[a-z]*|.[a-z]*\s.[a-z]*',i)
    condo = (df['neighbourhood'] == val) & (df['neighbourhood_group'].isnull())
    df.loc[condo,'neighbourhood_group'] = c[0]

In [170]:
# now we check if the neighbourhood_group column  is cleaned by applying lamda to the isnull function for full accuracy
df.apply(lambda x: sum(x.isnull()),axis=0)

name                                44
host_id                              0
host_name                            0
neighbourhood_group                  0
neighbourhood                        0
latitude                             0
longitude                            0
room_type                           17
price                                0
minimum_nights                      16
number_of_reviews                    0
last_review                       2758
reviews_per_month                 2758
calculated_host_listings_count       0
availability_365                   386
dtype: int64

thus we succeded in cleaning both Neighbourhood and neighbourhood_group column

# Replace missing values in minimum nights with the mean

In [171]:
# we shall replace missing values using the mean
# by first converting the dataframe to numeric data
df['minimum_nights'] = pd.to_numeric(df['minimum_nights'], errors= 'coerce')

In [172]:
# we now calculate the mean
df['minimum_nights'].mean()

17.530299188640974

In [173]:
# we now fill the mean of 17.53 calculated above in the missing values
df['minimum_nights'].fillna(17.53, inplace = True)

In [174]:
# lets now test if the null values have been filled
df.apply(lambda x: sum(x.isnull()),axis=0)

name                                44
host_id                              0
host_name                            0
neighbourhood_group                  0
neighbourhood                        0
latitude                             0
longitude                            0
room_type                           17
price                                0
minimum_nights                       0
number_of_reviews                    0
last_review                       2758
reviews_per_month                 2758
calculated_host_listings_count       0
availability_365                   386
dtype: int64

Thus as shown above we have successfully filled the remaining values of minimum nights with mean

# Explain how missing values are being replaced in hostname. Can you think of another solution for this?

The values in the host name are being replaced with the value 0 ti avoid inaccuracies of replacing the hostname 
with the wrong variable.

There are other solutions available for this as well such as using mode to fill the null values with the most common
hostname, as the host_id being a unique identifier of each name holds the aunthentic verification of the hostname
and also using the related neighbourhood_group to fill the missing hostname as Hostname Francesca resides in
the north reigion with the host_id 266763, thus we can fill with either and still have relative accuracy but more so with 
host_id

# Explain how missing values are being replaced in reviews per month. Can you think of another solution for this?

Missing values in reviews per Month are being replaced using the mean, and another solutio for this would be using
the Mode as we since this a cetegorical variable based on qualititative insights, using the most common review would 
give us a better view of what most customers are thinking and work according to it in improvements

# Remove all missing values in the bnb file according to your own plan if needed, i.e. you can change the way for 
# replacing the missings in one or more columns. In this case, mention which changes you have made

In [175]:
# Lets check in which columns do missing Values remain 
df.isnull().sum()

name                                44
host_id                              0
host_name                            0
neighbourhood_group                  0
neighbourhood                        0
latitude                             0
longitude                            0
room_type                           17
price                                0
minimum_nights                       0
number_of_reviews                    0
last_review                       2758
reviews_per_month                 2758
calculated_host_listings_count       0
availability_365                   386
dtype: int64

In [176]:
# we shall first replace all missing values using mode as its not a unique value, but we cannot drop it cause it will
# create disrepancies in our dataset
df['name'].mode()

0    Luxury hostel with in-cabin locker - Single mixed
dtype: object

In [177]:
# thus we fill it using the mode of 'Luxury hostel with in-cabin locker - Single mixed
df['name'].fillna('Luxury hostel with in-cabin locker - Single mixed', inplace = True)

In [178]:
# now we shall test if that worked 
df.isnull().sum()

name                                 0
host_id                              0
host_name                            0
neighbourhood_group                  0
neighbourhood                        0
latitude                             0
longitude                            0
room_type                           17
price                                0
minimum_nights                       0
number_of_reviews                    0
last_review                       2758
reviews_per_month                 2758
calculated_host_listings_count       0
availability_365                   386
dtype: int64

Thus we succed in eliminting the null values in the 'name' column

# Room Type Column

In [179]:
df['room_type'].isnull().sum()

17

In [180]:
df['room_type'].value_counts()

Entire home/apt    4126
Private room       3368
Shared room         392
2                     1
3                     1
1                     1
4                     1
Name: room_type, dtype: int64

In [181]:
# we shall use groupby to check which room type matches with which neigbourhood_group to atleast know 
# we first convert them both to category variables so they can be correlated
df= df.astype({'room_type': 'category', 'neighbourhood_group': 'category' })


In [182]:
df.groupby(['room_type', 'neighbourhood_group']).size()

room_type        neighbourhood_group
1                Central Region            0
                 East Region               0
                 North Region              1
                 North-East Region         0
                 West Region               0
2                Central Region            1
                 East Region               0
                 North Region              0
                 North-East Region         0
                 West Region               0
3                Central Region            0
                 East Region               0
                 North Region              1
                 North-East Region         0
                 West Region               0
4                Central Region            0
                 East Region               1
                 North Region              0
                 North-East Region         0
                 West Region               0
Entire home/apt  Central Region         3733
                 E

In [183]:
print(df[df['room_type'].isnull()]['neighbourhood_group'])

2252       Central Region
2253    North-East Region
2254          East Region
2255       Central Region
2256    North-East Region
2257       Central Region
2258       Central Region
2259       Central Region
7560       Central Region
7561          West Region
7562       Central Region
7563       Central Region
7564       Central Region
7565       Central Region
7566       Central Region
7567       Central Region
7568          East Region
Name: neighbourhood_group, dtype: category
Categories (5, object): [Central Region, East Region, North Region, North-East Region, West Region]


In [184]:
# now we shall fill the missing values in room_type based on which region has which highest room_type
# we here fill the missing values of 1 with east region
c0 = (df['neighbourhood_group'] == 'East Region') & (df['room_type'].isnull())
df.loc[c0, 'room_type'] = 'Private room'

c1 = (df['neighbourhood_group'] == 'Central Region') & (df['room_type'].isnull())
df.loc[c1, 'room_type'] = 'Entire home/apt'


c2 = (df['neighbourhood_group'] == 'North Region') & (df['room_type'].isnull())
df.loc[c2, 'room_type'] = 'Entire home/apt'


c3 = (df['neighbourhood_group'] == 'West Region') & (df['room_type'].isnull())
df.loc[c3, 'room_type'] = 'Shared room'

c4 = (df['neighbourhood_group'] == 'North-East Region') & (df['room_type'].isnull())
df.loc[c3, 'room_type'] = 'Private room'

In [185]:
# now we shall test if all null values have been filled in room_type
df.apply(lambda x: sum(x.isnull()))

name                                 0
host_id                              0
host_name                            0
neighbourhood_group                  0
neighbourhood                        0
latitude                             0
longitude                            0
room_type                            2
price                                0
minimum_nights                       0
number_of_reviews                    0
last_review                       2758
reviews_per_month                 2758
calculated_host_listings_count       0
availability_365                   386
dtype: int64

now we are only left with last_review and reviews_per_month and availability 365


# last_review

In [186]:
# we can replace all the missing values using mode as it will place the most common review review
# in the missing value
df['last_review'].mode

<bound method Series.mode of 0       21/10/2013
1       26/12/2014
2        1/10/2015
3        11/8/2019
4       28/07/2019
           ...    
7902           NaN
7903           NaN
7904           NaN
7905           NaN
7906           NaN
Name: last_review, Length: 7907, dtype: object>

In [187]:
df.last_review.mode()

0    12/8/2019
dtype: object

In [188]:
# thus we need to place [0] after mode using the fillna function to ensure it places the actual date as the mode
df['last_review'].fillna(df['last_review'].mode()[0],inplace = True)

In [189]:
# now we test if that succeeded
df.apply(lambda x: sum(x.isnull()))

name                                 0
host_id                              0
host_name                            0
neighbourhood_group                  0
neighbourhood                        0
latitude                             0
longitude                            0
room_type                            2
price                                0
minimum_nights                       0
number_of_reviews                    0
last_review                          0
reviews_per_month                 2758
calculated_host_listings_count       0
availability_365                   386
dtype: int64

now we are left with the null values of reviews_per_month and availibility_365

In [190]:
df

name    host_id  host_name  \
0                     COZICOMFORT LONG TERM STAY ROOM 2     266763  Francesca   
1                       Pleasant Room along Bukit Timah     227796    Sujatha   
2                                           COZICOMFORT     266763  Francesca   
3                   Ensuite Room (Room 1 & 2) near EXPO     367042    Belinda   
4                       B&B  Room 1 near Airport & EXPO     367042    Belinda   
...                                                 ...        ...        ...   
7902  Loft 2 pax near Haw Par / Pasir Panjang. Free ...  278109833      Belle   
7903                         3bedroom luxury at Orchard  238891646       Neha   
7904    [ Farrer Park ] New City Fringe CBD Mins to MRT  281448565      Mindy   
7905          Cheap Master Room in Central of Singapore  243835202      Huang   
7906  Amazing room with private bathroom walk to Orc...   28788520    Terence   

     neighbourhood_group neighbourhood  latitude  longitude        room_type  \
0           North Region     Woodlands   1.44255  103.79580                1   
1         Central Region   Bukit Timah   1.33235  103.78521                2   
2           North Region     Woodlands   1.44246  103.79667                3   
3            East Region      Tampines   1.34541  103.95712                4   
4            East Region      Tampines   1.34567  103.95963     Private room   
...                  ...           ...       ...        ...              ...   
7902      Central Region    Queenstown   1.27973  103.78751  Entire home/apt   
7903      Central Region       Tanglin   1.29269  103.82623  Entire home/apt   
7904      Central Region       Kallang   1.31286  103.85996     Private room   
7905      Central Region  River Valley   1.29543  103.83801     Private room   
7906      Central Region  River Valley   1.29672  103.83325     Private room   

      price  minimum_nights  number_of_reviews last_review  reviews_per_month  \
0        83          180.00                  1  21/10/2013               0.01   
1        81           90.00                 18  26/12/2014               0.28   
2        69            6.00                 20   1/10/2015               0.20   
3       206            1.00                 14   11/8/2019               0.15   
4        94            1.00                 22  28/07/2019               0.22   
...     ...             ...                ...         ...                ...   
7902    100           17.53                  0   12/8/2019                NaN   
7903    550           17.53                  0   12/8/2019                NaN   
7904     58           17.53                  0   12/8/2019                NaN   
7905     56           17.53                  0   12/8/2019                NaN   
7906   1500           17.53                  0   12/8/2019                NaN   

      calculated_host_listings_count  availability_365  
0                                  2             365.0  
1                                  1             365.0  
2                                  2             365.0  
3                                  9             353.0  
4                                  9             355.0  
...                              ...               ...  
7902                              31              61.0  
7903                              34             365.0  
7904                               3             173.0  
7905                               2              30.0  
7906                               7             365.0  

[7907 rows x 15 columns]

# availibility_365

In [191]:
# we apply mean here to fill the null values , since its only plausible to know Availibility on 
# average in particular durations
df['availability_365'].mean

<bound method Series.mean of 0       365.0
1       365.0
2       365.0
3       353.0
4       355.0
        ...  
7902     61.0
7903    365.0
7904    173.0
7905     30.0
7906    365.0
Name: availability_365, Length: 7907, dtype: float64>

In [192]:
df['availability_365'].fillna(df['availability_365'].mean(), inplace = True)

In [193]:
# now we test if the null values have been filled in availibility_365
df.apply(lambda s: sum(s.isnull()))

name                                 0
host_id                              0
host_name                            0
neighbourhood_group                  0
neighbourhood                        0
latitude                             0
longitude                            0
room_type                            2
price                                0
minimum_nights                       0
number_of_reviews                    0
last_review                          0
reviews_per_month                 2758
calculated_host_listings_count       0
availability_365                     0
dtype: int64

 thus we now only have reviews_per_month left 

# reviews_per_month

In [194]:
# we shall use mode here, as it would deem optimal to use the most common review by most customers to fill in 
# the missing values without denting accuracy by a large extent

df['reviews_per_month'].mode()

0    1.0
dtype: float64

In [195]:
# thus we we fill in the mode within the missing values 
df['reviews_per_month'].fillna(df['reviews_per_month'].mode()[0], inplace = True)

In [196]:
# now we confirm if our dataset is clean
df.apply(lambda b: sum(b.isnull()))

name                              0
host_id                           0
host_name                         0
neighbourhood_group               0
neighbourhood                     0
latitude                          0
longitude                         0
room_type                         2
price                             0
minimum_nights                    0
number_of_reviews                 0
last_review                       0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
dtype: int64

In [197]:
# we replace the remaining null values of room_type with mode
df['room_type'].mode

<bound method Series.mode of 0                     1
1                     2
2                     3
3                     4
4          Private room
             ...       
7902    Entire home/apt
7903    Entire home/apt
7904       Private room
7905       Private room
7906       Private room
Name: room_type, Length: 7907, dtype: category
Categories (7, object): [1, 2, 3, 4, Entire home/apt, Private room, Shared room]>

In [198]:
df['room_type'].fillna(df['room_type'].mode()[0], inplace= True)

In [199]:
# now we confirm if our dataset is clean
df.apply(lambda b: sum(b.isnull()))

name                              0
host_id                           0
host_name                         0
neighbourhood_group               0
neighbourhood                     0
latitude                          0
longitude                         0
room_type                         0
price                             0
minimum_nights                    0
number_of_reviews                 0
last_review                       0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
dtype: int64

In [200]:
# now we conduct ANOVA and T-test to check the accuracy of our results of cleaning methods 
# we first import the statistics library for it and import stats module from scipy

In [201]:
import statistics as stat
from scipy import stats

# Conduct t-test between the relevant columns and explain the result (it should make logical sense to conduct the test)

In [202]:
# first we shall conduct the t-test between 
# number_of_reviews and minimum_nights to test how much customers like spendinf time at each lodge and do more 
# nights being spent result in less satisfaction for customer
df['number_of_reviews'].astype(int)
df['minimum_nights'].astype(int)

0       180
1        90
2         6
3         1
4         1
       ... 
7902     17
7903     17
7904     17
7905     17
7906     17
Name: minimum_nights, Length: 7907, dtype: int32

In [203]:
test = stats.ttest_ind(df['number_of_reviews'], df['minimum_nights'])

In [204]:
print(test)

Ttest_indResult(statistic=-8.153013932202944, pvalue=3.8136852863122826e-16)


 since the value is greater than 0.05 thus we fail to reject our null hypothesis that there is no difference
 and number of reviews and minimum_nights dont have a strong

In [205]:
# we now check it for minimum_nights and availibility_365
test1 = stats.ttest_ind(df['minimum_nights'], df['availability_365'])
print(test1)

Ttest_indResult(statistic=-115.41578866935224, pvalue=0.0)


 thus since here the value is less than 0.05 thus we can reject our null hypothesis of there being no difference 
 and that minimum_nights and availibility_365 have a relationship as more customers spending nights does affect 
 room availibility

In [216]:
# we now check it for price and host_id
test2 = stats.ttest_ind(df['price'], df['minimum_nights'])
print(test2)

Ttest_indResult(statistic=39.323322221931264, pvalue=0.0)


thus with probability being less than 0.05 we reject our null hypothesis here, showing there is a relationship 
between price and how much customers are willing to stay on.

#   Conduct ANOVA (one-way, i.e., one string and one numerical column) between the relevant columns and explain the result (it should make logical sense to conduct the test)

In [219]:
# we first create a list of all num_columns for ease of calculation
num_cols = ['latitude','longitude','price','minimum_nights','number_of_reviews','reviews_per_month','calculated_host_listings_count','availability_365','Total_Amount_Paid']

In [228]:
# we convert last_review to datetime to use it in the ANOVA test
df['last_review'] = pd.to_datetime(df['last_review'])

In [233]:
# now we start the ANOVA one_way test between 'last_review' and 'review_per_month'
ANOVA_test = stats.f_oneway(df['last_review'],df['reviews_per_month'])

In [234]:
print(ANOVA_test)

F_onewayResult(statistic=24922358.898263037, pvalue=0.0)


Thus having a p < 0.05 proves that we can reject our null hypothesis and prove with a 95% assurity 
that a difference in means exist betweem last_review and review_per_month thus making it hard for Airbnb stakeholders 
to deny that 'we are not receieving many customer reviews'

In [237]:
# now we start the ANOVA one_way test between 'minimum_nights' and 'availibility_365'
ANOVA_test2 = stats.f_oneway(df['minimum_nights'],df['availability_365'])

In [238]:
print(ANOVA_test2)

F_onewayResult(statistic=13320.804274168559, pvalue=0.0)


Thus having a p < 0.05 proves that we can reject our null hypothesis and prove with a 95% assurity that a difference in means exist between minimum_nights and availability_365 thus making it easier for Airbnbs management to claim that customers spedning more nights is affecting space available for more customers throughout the year.

In [243]:
# now we start the ANOVA one_way test between 'minimum_nights' and 'availibility_365'
ANOVA_test3 = stats.f_oneway(df['price'],df['calculated_host_listings_count'])

In [244]:
print(ANOVA_test3)

F_onewayResult(statistic=1093.3273786192103, pvalue=6.790458153231057e-232)


Over here we fail to reject our null hypothesis as p > 0.05 proves the means are different betweem prices charged and calculated host lisitings. Thus proving that some hosts may be charging the same prices.